crates/ironrdp-rdpdr/src/pdu/esc.rs

```diff
@@ -787,483 +886,202 @@ impl rpce::HeaderlessEncode for GetStatusChangeReturn {
-        version: u8,
-        endianness: Endianness,
-        common_header_length: u16,
-        filler: u32,
-    }
-
-    impl Default for StreamHeader {
-        fn default() -> Self {
-            Self {
-                version: 1,
-                endianness: Endianness::LittleEndian,
-                common_header_length: 8,
-                filler: 0xCCCC_CCCC,
-            }
-        }
-    }
-
-    impl StreamHeader {
-        const NAME: &'static str = "RpceStreamHeader";
-
-        fn encode(&self, dst: &mut WriteCursor<'_>) -> PduResult<()> {
-            ensure_size!(in: dst, size: Self::size());
-            dst.write_u8(self.version);
-            dst.write_u8(self.endianness.into());
-            dst.write_u16(self.common_header_length);
-            dst.write_u32(self.filler);
-            Ok(())
-        }
-
-        fn decode(src: &mut ReadCursor<'_>) -> PduResult<Self> {
-            ensure_size!(in: src, size: Self::size());
-            let version = src.read_u8();
-            let endianness = Endianness::try_from(src.read_u8())?;
-            let common_header_length = src.read_u16();
-            let filler = src.read_u32();
-            if endianness == Endianness::LittleEndian {
-                Ok(Self {
-                    version,
-                    endianness,
-                    common_header_length,
-                    filler,
-                })
-            } else {
-                Err(invalid_message_err!(
-                    "decode",
-                    "StreamHeader",
-                    "server returned big-endian data, parsing not implemented"
-                ))
-            }
-        }
-
-        fn size() -> usize {
-            size_of::<u8>() + size_of::<u8>() + size_of::<u16>() + size_of::<u32>()
-        }
-    }
-
-    #[derive(Debug, Clone, Copy, PartialEq)]
-    #[repr(u8)]
-    enum Endianness {
-        BigEndian = 0x00,
-        LittleEndian = 0x10,
+    /// [2.2.5 Protocol Identifier]: https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpesc/41673567-2710-4e86-be87-7b6f46fe10af
+    #[derive(Debug, Clone)]
+    pub struct CardProtocol: u32 {
+        const SCARD_PROTOCOL_UNDEFINED = 0x00000000;
+        const SCARD_PROTOCOL_T0 = 0x00000001;
+        const SCARD_PROTOCOL_T1 = 0x00000002;
+        const SCARD_PROTOCOL_TX = 0x00000003;
+        const SCARD_PROTOCOL_RAW = 0x00010000;
+        const SCARD_PROTOCOL_DEFAULT = 0x80000000;
+        const SCARD_PROTOCOL_OPTIMAL = 0x00000000;
     }
+}
 
-    impl TryFrom<u8> for Endianness {
-        type Error = PduError;
+/// [2.2.1.2 REDIR_SCARDHANDLE]
+///
+/// [2.2.1.2 REDIR_SCARDHANDLE]: https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpesc/b6276356-7c5f-4d3e-be92-a6c85e58d008
+#[derive(Debug)]
+pub struct ScardHandle {
+    context: ScardContext,
+    length: u32,
+    // Shortcut: we always create 4-byte handle values.
+    // The spec allows this field to have variable length.
+    value: u32,
```

Comment:
Not a big deal, but since we have an hardcoded length of 4, you may consider to not hold the `length` field at all and use a constant instead.

```rust
pub struct ScardHandle {
    context: ScardContext,
    // Shortcut: we always create 4-byte handle values.
    // The spec allows this field to have variable length.
    value: u32,
}

impl ScardHandle {
    const NAME: &'static str = "REDIR_SCARDHANDLE";
    const VALUE_LENGTH: u32 = 4; // <-

    /* … */
}

impl ndr::Decode for ScardHandle {
    fn decode_ptr(src: &mut ReadCursor<'_>, index: &mut u32) -> PduResult<Self>
    where
        Self: Sized,
    {
        let context = ScardContext::decode_ptr(src, index)?;
        ensure_size!(ctx: "Handle::decode_ptr", in: src, size: size_of::<u32>());
        let length = src.read_u32();
        if length != Self::VALUE_LENGTH { // <-
            todo!("return an error, reporting unsupported value length");
        }
        let _ptr = ndr::decode_ptr(src, index)?;
        Ok(Self {
            context,
            value: 0,
        })
    }

    fn decode_value(&mut self, src: &mut ReadCursor<'_>) -> PduResult<()> {
        self.context.decode_value(src)?;
        ensure_size!(in: src, size: size_of::<u32>());
        let length = src.read_u32();
        if length != Self::VALUE_LENGTH { // <-
            Err(invalid_message_err!(
                "decode_value",
                "mismatched length in Handle reference and value"
            ))
        } else {
            ensure_size!(in: src, size: size_of::<u32>());
            self.value = src.read_u32();
            Ok(())
        }
    }
}

impl ndr::Encode for ScardHandle {
    /* … */

    fn encode_value(&self, dst: &mut WriteCursor<'_>) -> PduResult<()> {
        ensure_size!(in: dst, size: self.size_value());
        self.context.encode_value(dst)?;
        dst.write_u32(Self::VALUE_LENGTH); // <-
        dst.write_u32(self.value);
        Ok(())
    }

    /* … */
}
```

I think it’s easier to reason about: since we fix the value to `u32`, we also fix the length field to `4`.

(Of course, if supporting variable-length values is a thing we’ll need in the future, you may want to keep the current code.)

Note: I’ve seen something similar a several places, you may want to reflect this everywhere.

Reply:
IMO its preferable to keep it as is: currently, if the server happens to send `ScardContext` or `ScardHandle` of a different length, we can handle that, whereas with the proposed changes the program would fail. Granted it's unlikely that a server sends us back anything other than exactly what we sent it, but I don't see good reason to make changes that make this more fragile.

Reply:
I agree with the rationale, but I’m not sure to understand how the current implement is actually more robust: is it actually handling payloads bigger or smaller than 4 bytes? For example, at [this place](https://github.com/Devolutions/IronRDP/blob/464f154ea0558133b556af541ee35f342335abd3/crates/ironrdp-rdpdr/src/pdu/esc.rs#L108-L117):
```rust
fn decode_value(&mut self, src: &mut ReadCursor<'_>) -> PduResult<()> {
    ensure_size!(in: src, size: size_of::<u32>()*2);
    let length = src.read_u32();
    if length != self.length {
        error!(?length, ?self.length, "Mismatched length in ScardContext reference and value");
        Err(invalid_message_err!(
            "decode_value",
            "mismatched length in ScardContext reference and value"
        ))
    } else {
        self.value = src.read_u32();
        Ok(())
    }
}
```
I understand that if length is not 4, then `src.read_u32()` is basically reading either not enough or too much, which is just a bug instead of an explicit error.

Reply:
Ah true, mistake on my part.

Reply:
d978bf9cb19b6dc606aef23dd528f50e993e767d

Reply:
Thank you!